pol-core/pol/uoexec.cpp

```diff
@@ -750,16 +750,58 @@ bool UOExecutor::getObjtypeParam( unsigned param, const Items::ItemDesc*& itemde
 
 bool UOExecutor::getSkillIdParam( unsigned param, USKILLID& skillid )
 {
-  int skillval;
-  if ( getParam( param, skillval, SKILLID__LOWEST, networkManager.uoclient_general.maxskills ) )
+  BObjectImp* imp = getParamImp( param );
+  if ( imp == nullptr )
   {
-    skillid = static_cast<USKILLID>( skillval );
-    return true;
+    setFunctionResult( new BError( "Missing parameter " + Clib::tostring( param ) ) );
+    return false;
   }
-  else
+  else if ( imp->isa( BObjectImp::OTLong ) )
   {
+    BLong* plong = Clib::explicit_cast<BLong*, BObjectImp*>( imp );
+    int value = plong->value();
+    if ( value >= SKILLID__LOWEST && value <= networkManager.uoclient_general.maxskills )
+    {
+      skillid = static_cast<USKILLID>( value );
+      return true;
+    }
+    else
+    {
+      std::string report = "Parameter " + Clib::tostring( param ) + " value " +
+                           Clib::tostring( value ) + " out of expected range of [" +
+                           Clib::tostring( SKILLID__LOWEST ) + ".." +
+                           Clib::tostring( networkManager.uoclient_general.maxskills ) + "]";
+      setFunctionResult( new BError( report ) );
+      return false;
+    }
+  }
+  else if ( imp->isa( BObjectImp::OTString ) )
+  {
+    const Mobile::Attribute* attr;
+    if ( !getAttributeParam( param, attr ) )
+      return false;
+    for ( unsigned short i = 0; i <= networkManager.uoclient_general.maxskills; ++i )
+    {
+      const UOSkill& uoskill = GetUOSkill( i );
+      if ( uoskill.pAttr == attr )
+      {
+        skillid = static_cast<USKILLID>( uoskill.skillid );
+        return true;
+      }
+    }
```

Comment:
Could we do this lookup once when loading the attribute data (set on a field on Attribute), rather than on every getSkillIdParam (by attrname) call?

Reply:
Yes true, it shouldn't really matter since not many attributes exists, but it also doesn't hurt to store the extra information.